Add OpenGL renderer

Calinou opened the pull request:
Based on @lawnjelly's efforts on the GLES2 renderer here: https://github.com/godotengine/godot/pull/44399 :slightly_smiling_face:

Despite the [news article stating that OpenGL support will return in Godot 4.1](https://godotengine.org/article/about-godot4-vulkan-gles3-and-gles2), this pull request is targeting a merge for Godot 4.0. However, **we don't expect this OpenGL renderer to be fully complete (especially on the 3D side) when Godot 4.0 is released**. Nonetheless, it may still be sufficient for many 2D games and non-game applications that want to maximize compatibility.

Commits are unsquashed for now to ease troubleshooting and tracking regressions. Most of them should be squashed before merging this PR.

This closes https://github.com/godotengine/godot-proposals/issues/877.

### OpenGL renderer's advantages over Vulkan

- Much better compatibility with old devices (desktop, mobile).
- Can run on HTML5 using WebGL 2.0 once the platform layer is reimplemented.
  - In contrast, [WebGPU isn't available in all stable browsers yet](https://caniuse.com/webgpu). Moreover, a RenderingDevice abstraction needs to be written for it.
- Faster startup times thanks to simpler shaders (even with the lack of shader caching).

*When targeting desktop platforms, using Vulkan is still recommended for several reasons:*

### Vulkan renderer's advantages over OpenGL

- Better visuals and performance in complex scenes.
  - If you need a lower base cost for integrated graphics or low-end dedicated GPUs, this can be achieved by using the Vulkan Mobile backend on desktop platforms.
- No issues with AMD GPU performance on Windows (due to subpar OpenGL implementation).

Nonetheless, for mobile platforms, I expect OpenGL to be more prevalent for a while.

### Differences from [lawnjelly's pull request](https://github.com/godotengine/godot/pull/44399)

- OpenGL 3.3 / OpenGL ES 3.0 / WebGL 2.0 is now used as a baseline. **The renderer will still follow a low-end-friendly approach** to maximize performance and compatibility on low-end devices.
- Removed the `gles_common` abstraction as there will only be one OpenGL renderer.
- "GLES2" renderer renamed to "OpenGL" for consistency with the Vulkan renderer, which does not specify a version in its name. Use `--rendering-driver opengl` to force a project to start with the OpenGL renderer, or use the dropdown in the top-right corner of the editor.

### Current limitations

- **2D only.** No 3D rendering support for now.
  - Getting 2D rendering support to be reliable for daily use is the current priority, as it'll unblock contributors who don't have a GPU that supports Vulkan.
- Only the Linux/X11 context is currently implemented. Due to this, the OpenGL renderer can't be used on Windows, macOS, Android, iOS and HTML5 yet.

### TODO

- 2D editor viewport rendering is currently broken, but the running project displays correctly.
- The OpenGL Core profile should be used instead of the Compatibility profile. The compatibility profile was temporarily enabled to allow Godot to render correctly after switching from a GLES2 context to a GLES3 context.
- Shaders should be ported from GLSL ES 1.0 to GLSL ES 3.0 using the Vulkan GLSL shaders as a base.
- Some texture functions need to be reimplemented.
- The buildsystem code should be checked to make sure OpenGL and Vulkan can be disabled individually at compile-time. This lets people build smaller binaries when needed.
- The shader compiler needs to be rewritten to use a modern approach. Ubershaders and asynchronous compilation will likely be used to avoid stutters during gameplay. This will be handled by @reduz and @RandomShaper.
- Decide what to do with the `GLES_ENABLED` and `OPENGL_BACKEND_ENABLED` defines. Should we merge them with `OPENGL_ENABLED` to be consistent with `VULKAN_ENABLED`?
- rasterizer_array.h should be replaced with LocalVector: https://github.com/Calinou/godot/blob/add-opengl-renderer/drivers/opengl/rasterizer_array.h#L37

## Preview

![Godot editor running on OpenGL](https://user-images.githubusercontent.com/180032/134987888-38e5e1ba-faa4-4b9f-8f0f-411f97bc7469.png)

lawnjelly committed:
Godot 4 GLES2 2D renderer + linux display manager

First implementation with linux display manager
Co-authored-by:clayjohn <claynjohn@gmail.com>
Co-authored-by:Fabio Alessandrelli <fabio.alessandrelli@gmail.com>

lawnjelly committed:
Added GLES2 vsync support

lawnjelly committed:
GLES2 tidying, TIME and bootscreen

lawnjelly committed:
Add single threaded mode for EditorResourcePreview

lawnjelly committed:
Fixes for quarmin's worst possible project.

lawnjelly committed:
Rebasing and fixes to compile after a year gap.

Calinou committed:
Fix shader builder code generation for the latest `master` branch

Calinou committed:
Remove references to a nonexistent GLES3 renderer for now [ci skip]

Calinou committed:
Remove unused/deprecated code and references to Godot 3.x [ci skip]

Calinou committed:
Merge the `gles_common/` abstraction folder with `gles2/` [ci skip]

Calinou committed:
Rename GLES2 driver to OpenGL [ci skip]

This is consistent with the Vulkan driver which does not specify
a version within the driver name.

Moreover, this avoids confusion with different OpenGL versions being
used depending on the platform (OpenGL on desktop, OpenGL ES on mobile).

Calinou committed:
[WIP] Initialize OpenGL 3.3 / OpenGL ES 3.0 context [ci skip]

Shaders don't compile yet.

Calinou committed:
Port 2D shader to GLSL ES 3.0 to fix shader compilation [ci skip]

This does not fix rendering due to incorrect OpenGL usage.

Calinou committed:
Revert "Port 2D shader to GLSL ES 3.0 to fix shader compilation [ci skip]"

We'll try to use the GLSL ES 1.0 shaders directly for now.

This reverts commit c945b56b41655f815df92025479c0713397365fe.

Calinou committed:
Update some more references to OpenGL

Calinou committed:
Support PI, TAU and E shader constants in OpenGL shaders

Calinou committed:
Use OpenGL compatibility profile to fix rendering for now

The renderer currently relies on deprecated OpenGL APIs.

Calinou committed:
Fix code style issue reported by Black

Calinou committed:
Update the VideoDriver enum value for the OpenGL renderer

- Move Vulkan to the top of the enum to be consistent with the editor
  dropdown and project setting.

Calinou committed:
Rename all references to GLES2 to OpenGL

- Remove some unused GLES3 builder code.

clayjohn commented:
A couple notes:
1. Have you tested on Windows? I had GLES3 working in Windows on Lawnjelly's branch earlier, I imagine that it should be working or close to working still
2. The rasterizer should not depend on the dummy rasterizer at all. For now, it should be enough to copy-paste the relevant sections of the dummy into the relevant files in drivers/GLES3

aaronfranke commented on core/core_bind.h:
```diff
@@ -119,8 +119,8 @@ class OS : public Object {
 	enum VideoDriver {
-		VIDEO_DRIVER_GLES2,
 		VIDEO_DRIVER_VULKAN,
+		VIDEO_DRIVER_OPENGL,
```
`O` comes before `V` in the alphabet, there's no purpose to swapping the order here.

Calinou replied:
I swapped the order to match the order found in the project settings' enum: https://github.com/Calinou/godot/blob/8fb7cafaeb31b9d27b199effbc8a1829a0fa4845/main/main.cpp#L1271

qarmin commented:
I think that it should be temporary to CI to see progress by adding `--rendering-driver opengl` to this commands
https://github.com/godotengine/godot/blob/8abd46096ae52025d54bf58a782a6c057b3254f6/.github/workflows/linux_builds.yml#L115-L127

qarmin commented:
On Intel HD 3000 and Mesa 21.3, I see empty screen in both project manager and editor
![Zrzut ekranu z 2021-09-28 08-46-27](https://user-images.githubusercontent.com/41945903/135037125-69a38fab-6dd3-49c6-929f-0b4ed65b3796.png)
and also I see this errors
```
Mesa: User error: GL_INVALID_OPERATION in glDrawArrays
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glDrawArrays
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glDrawElements
Mesa: User error: GL_INVALID_OPERATION in glVertexAttribPointer(no array object bound)
Mesa: User error: GL_INVALID_OPERATION in glDrawArrays
Mesa: User error: GL_INVALID_VALUE in glTexImage2D(internalFormat=GL_LUMINANCE_ALPHA)
Mesa: User error: GL_INVALID_VALUE in glTexImage2D(internalFormat=GL_LUMINANCE_ALPHA)
```

Calinou committed:
Replace use of `#pragma once`, remove commented out code in stubs

Calinou committed:
Remove reliance on the Dummy rasterizers

jordo commented:
Is Godot 4.0 launching without html5/web support then?

Calinou committed:
Always accept title-case driver names to make the OpenGL project setting work

- Mark the driver name setting as requiring a project restart.

Calinou commented:
> Is Godot 4.0 launching without html5/web support then?

It depends on whether we can finish this OpenGL renderer (+ the WebGL platform port) in time. Either way, `3.x` will remain supported until 4.1 is released.

jordo commented:
> It depends on whether we can finish this OpenGL renderer (+ the WebGL platform port) in time for 4.0's release. We can't promise anything yet. Either way, `3.x` will remain supported until 4.1 is released.
> 
> WebGPU isn't likely to be implemented until 4.1 at least, so OpenGL via WebGL is our only option for HTML5 in the short term.


Would love a concise answer.  I understand 3.x is supported, but my question was specific to 4.0.  I'm not looking for promises, it would just be nice to be clear about this so developers can plan accordingly.  From other contributer comments, it seems Vulkan is going to be the only supported initial backend for 4.0.  Connecting the dots would therefore mean that web export won't be supported in 4.0 correct?  WebGPU isn't a viable option right now, nor the relatively near future.

I think this is an important platform to confirm this one way or the other.

jordo commented:
I guess the PR of the actual renderer impl isn't the best place to ask these questions, but it would be nice to know if the plan is to have this available in 4.0 or not.  It's fine either way, just nice to know so people can plan and adjust accordingly.

Calinou commented:
> Connecting the dots would therefore mean that web export won't be supported in 4.0 correct?

Indeed, web export (and web editor) won't be available until some way to render graphics is implemented in the `master` branch.

Calinou committed:
Rename `GLES_<platform>_ENABLED` defines to `OPENGL_<platform>_ENABLED`

This also enables building OpenGL for Windows.

Calinou committed:
Replace platform-specific OpenGL defines with `OPENGL_ENABLED`

This is consistent with `VULKAN_ENABLED`.

Calinou commented:
> 1. Have you tested on Windows? I had GLES3 working in Windows on Lawnjelly's branch earlier, I imagine that it should be working or close to working still

It seems to crash on startup, but Vulkan also crashes in the same branch anyway. I haven't investigated why yet.

clayjohn commented:
This is a good start! I was able to get it running on Windows (PR here https://github.com/Calinou/godot/pull/6)

A couple of things we need to take a look at still:
1. It looks like viewports are not working
2. The renderer is still using deprecated functionality not available in OpenGL 3.3, we should transition to OpenGL 3.3 fully before merging (can can likely copy-paste a lot from the GLES3 drivers)
3. Multiple windows are not working (and single window mode is broken because of 1)

clayjohn committed:
Fix OpenGL initialization on Windows

Calinou committed:
Update for latest `master` branch

clayjohn commented:
Okay, I've submitted a pull request that removes much of the deprecated GLES2 functions (enough to get core profile working anyway). I have also fixed the bug that made viewports not display. https://github.com/Calinou/godot/pull/7

So right now we have a minimal 2D renderer in OpenGL 3.3 on Windows. @Calinou If you can set the X11 platform to core profile and verify that it works there as well then I think we have a minimal working example here!

clayjohn committed:
Use OpenGL 3.3 Core Profile (#7)

Calinou committed:
Remove outdated comment about OpenGL Compatibility Profile

Calinou committed:
Fix build error after editor toast error implementation

Calinou committed:
Hide renderer selection dropdown until OpenGL support is more mature

The renderer can still be changed in the Project Settings or using
the `--rendering-driver opengl` command line argument.

Calinou committed:
Remove commented out exporter code

Calinou committed:
Remove some OpenGL/DisplayServer-related debugging prints
